In [2]:
# importing libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [3]:
# defining the path
path = r'C:\Users\dog\Desktop\Instacart Basket Analysis'


In [4]:
# importing customers DF
df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data','customers.csv'))

In [5]:
df_customers.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [6]:
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [7]:
df_customers.shape

(206209, 10)

In [8]:
df_customers.head(5)

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [9]:
# renaming columns for better readability
df_customers = df_customers.rename(columns={'First Name': 'first_name','Surnam': 'surname','Gender': 'gender', 'STATE': 'state', 'Age': 'age', 'date_joined' :'registration_date', 'n_dependants' : 'number_dependants', 'fam_status' : 'marital_status'})

In [10]:
# checking if that worked out
df_customers.head(5)

user_id first_name   surname  gender       state  age registration_date  \
0    26711    Deborah  Esquivel  Female    Missouri   48          1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36          1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35          1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40          1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26          1/1/2017   

   number_dependants marital_status  income  
0                  3        married  165665  
1                  0         single   59285  
2                  2        married   99568  
3                  0         single   42049  
4                  1        married   40374

In [11]:
# Finding values with null or NaN values
mask = df_customers.isnull().any(axis=1)

In [12]:
# checking which columns has NaN or null
df_customers[mask].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11259 entries, 53 to 206171
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   user_id            11259 non-null  int64 
 1   first_name         0 non-null      object
 2   surname            11259 non-null  object
 3   gender             11259 non-null  object
 4   state              11259 non-null  object
 5   age                11259 non-null  int64 
 6   registration_date  11259 non-null  object
 7   number_dependants  11259 non-null  int64 
 8   marital_status     11259 non-null  object
 9   income             11259 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 967.6+ KB


####  there are 11259 rows where first name is missing. since any other data is included, there is nothing needed to be done with that

In [41]:
# checking for the dups in the data frame
df_customers_dups = df_customers.duplicated()

In [45]:
df_customers_dups.info()

<class 'pandas.core.series.Series'>
RangeIndex: 206209 entries, 0 to 206208
Series name: None
Non-Null Count   Dtype
--------------   -----
206209 non-null  bool 
dtypes: bool(1)
memory usage: 201.5 KB


In [48]:
df_customers_dups.value_counts()

False    206209
dtype: int64

####  no duplicates are found

In [54]:
# checking for mixed data types
for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)

first_name


#### first_name column has mixed format

In [55]:
# changing column type to str
df_customers['first_name'] = df_customers['first_name'].astype('str')

In [62]:
# merging the customers df with the most recent df_ords_prods_merged df
#importing df_ords_prods_merged df
df_ords_prods_merged_new = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_new02.01.pkl'))

In [64]:
# checking if the df was imported and if that is the right one
df_ords_prods_merged_new.shape

(32404859, 26)

In [65]:
df_ords_prods_merged_new.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  ...  \
0                     NaN         196                  1          0  ...   
1                    15.0         196                  1          1  ...   
2                    21.0         196                  1          1  ...   
3                    29.0         196                  1          1  ...   
4                    28.0         196                  1          1  ...   

      price_range_loc     busiest_day  busiest_days_new   busiest_hours  \
0  High-range product  Regularly busy    Regularly busy  average orders   
1  High-range product  Regularly busy      Slowest days  average orders   
2  High-range product  Regularly busy      Slowest days    Most orders    
3  High-range product      Least busy      Slowest days  average orders   
4  High-range product      Least busy      Slowest days    Most orders    

   max_orders customers_loyalty mean_price spending_characteristic  \
0          10      New customer   6.367797             Low spender   
1          10      New customer   6.367797             Low spender   
2          10      New customer   6.367797             Low spender   
3          10      New customer   6.367797             Low spender   
4          10      New customer   6.367797             Low spender   

  since_priors_order_median     ordering_frequency  
0                      20.5  Non-frequent customer  
1                      20.5  Non-frequent customer  
2                      20.5  Non-frequent customer  
3                      20.5  Non-frequent customer  
4                      20.5  Non-frequent customer  

[5 rows x 26 columns]

In [66]:
df_ords_prods_merged_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 26 columns):
 #   Column                     Dtype   
---  ------                     -----   
 0   order_id                   int64   
 1   user_id                    int64   
 2   eval_set                   object  
 3   order_number               int64   
 4   order_dow                  int64   
 5   order_hour_of_day          int64   
 6   days_since_prior_order     float64 
 7   product_id                 int64   
 8   add_to_cart_order          int64   
 9   reordered                  int64   
 10  Unnamed: 0                 int64   
 11  product_name               object  
 12  aisle_id                   int64   
 13  department_id              int64   
 14  prices                     float64 
 15  _merge                     category
 16  price_range_loc            object  
 17  busiest_day                object  
 18  busiest_days_new           object  
 19  busiest_hours      

In [67]:
df_ords_prods_merged_new.describe()

order_id       user_id  order_number     order_dow  \
count  3.240486e+07  3.240486e+07  3.240486e+07  3.240486e+07   
mean   1.710745e+06  1.029372e+05  1.714230e+01  2.738867e+00   
std    9.872988e+05  5.946610e+04  1.753532e+01  2.090077e+00   
min    2.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.559470e+05  5.142200e+04  5.000000e+00  1.000000e+00   
50%    1.711049e+06  1.026160e+05  1.100000e+01  3.000000e+00   
75%    2.565499e+06  1.543890e+05  2.400000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  9.900000e+01  6.000000e+00   

       order_hour_of_day  days_since_prior_order    product_id  \
count       3.240486e+07            3.032876e+07  3.240486e+07   
mean        1.342515e+01            1.110408e+01  2.559866e+04   
std         4.246380e+00            8.779064e+00  1.408400e+04   
min         0.000000e+00            0.000000e+00  1.000000e+00   
25%         1.000000e+01            5.000000e+00  1.354400e+04   
50%         1.300000e+01            8.000000e+00  2.530200e+04   
75%         1.600000e+01            1.500000e+01  3.794700e+04   
max         2.300000e+01            3.000000e+01  4.968800e+04   

       add_to_cart_order     reordered    Unnamed: 0      aisle_id  \
count       3.240486e+07  3.240486e+07  3.240486e+07  3.240486e+07   
mean        8.352547e+00  5.895873e-01  2.560037e+04  7.119612e+01   
std         7.127071e+00  4.919087e-01  1.408555e+04  3.821139e+01   
min         1.000000e+00  0.000000e+00  0.000000e+00  1.000000e+00   
25%         3.000000e+00  0.000000e+00  1.354400e+04  3.100000e+01   
50%         6.000000e+00  1.000000e+00  2.530300e+04  8.300000e+01   
75%         1.100000e+01  1.000000e+00  3.795100e+04  1.070000e+02   
max         1.450000e+02  1.000000e+00  4.969200e+04  1.340000e+02   

       department_id        prices    max_orders    mean_price  \
count   3.240486e+07  3.240486e+07  3.240486e+07  3.240486e+07   
mean    9.919792e+00  1.198023e+01  3.305217e+01  1.198023e+01   
std     6.281485e+00  4.956554e+02  2.515525e+01  8.324227e+01   
min     1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%     4.000000e+00  4.200000e+00  1.300000e+01  7.387298e+00   
50%     9.000000e+00  7.400000e+00  2.600000e+01  7.824786e+00   
75%     1.600000e+01  1.130000e+01  4.700000e+01  8.254023e+00   
max     2.100000e+01  9.999900e+04  9.900000e+01  2.500542e+04   

       since_priors_order_median  
count               3.240485e+07  
mean                1.039776e+01  
std                 7.131754e+00  
min                 0.000000e+00  
25%                 6.000000e+00  
50%                 8.000000e+00  
75%                 1.300000e+01  
max                 3.000000e+01

In [68]:
# using outer join so that we do not lose any data 
df_ords_prods_custmrs = pd.merge(df_ords_prods_merged_new, df_customers, on = 'user_id', how = 'outer')

In [69]:
# checking if the merge function worked successfully
df_ords_prods_custmrs.shape

(32404859, 35)

#### new columns was added and data frames were merged successfully 

In [70]:
# exorting file as a pickle
df_ords_prods_custmrs.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'df_ords_prods_custmrs.pkl'))